packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol

```diff
@@ -158,4 +735,366 @@ library ConnextUtils {
+    uint256 toSwap = _args.executeArgs.amount;
+    uint256 pathLen = _args.executeArgs.routers.length;
+    if (_isFast) {
+      // this is the fast liquidity path
+      // ensure the router is whitelisted
+
+      // calculate amount with fast liquidity fee
+      toSwap = _getFastTransferAmount(
+        _args.executeArgs.amount,
+        _args.LIQUIDITY_FEE_NUMERATOR,
+        _args.LIQUIDITY_FEE_DENOMINATOR
+      );
+
+      // TODO: validate routers signature on path / transferId
+
+      // store the routers address
+      _routedTransfers[_transferId] = _args.executeArgs.routers;
+
+      // for each router, assert they are approved, and deduct liquidity
+      uint256 routerAmount = toSwap / pathLen;
+      for (uint256 i; i < pathLen; ) {
+        // while theres no way for a router to have sufficient liquidity
+        // if they have never been approved, this check ensures they weren't
+        // removed from the whitelist
+        if (!_routerInfo.approvedRouters[_args.executeArgs.routers[i]]) {
+          revert ConnextUtils__execute_notSupportedRouter();
+        }
+
+        // decrement routers liquidity
+        _routerBalances[_args.executeArgs.routers[i]][_args.executeArgs.local] -= routerAmount;
+
+        unchecked {
+          i++;
+        }
+      }
+    }
+
+    // swap out of mad* asset into adopted asset if needed
+    return
+      _swapFromLocalAssetIfNeeded(
+        _canonicalToAdopted,
+        _adoptedToLocalPools,
+        _args.tokenRegistry,
+        _args.executeArgs.local,
+        toSwap
+      );
+  }
+
+  /**
+   * @notice Handles transferring funds from msg.sender to the Connext contract.
+   * @dev This overload is used to prevent stack too deep errors
+   * @return The amount of the asset that was seen by the contract (may not be the specifiedAmount
+   * if the token is a fee-on-transfer token)
+   */
+  function _handleIncomingAsset(xCallLibArgs calldata _args) private returns (address, uint256) {
+    return
+      _handleIncomingAsset(
+        _args.xCallArgs.transactingAssetId,
+        _args.xCallArgs.amount,
+        _args.xCallArgs.relayerFee,
+        _args.wrapper
+      );
+  }
+
+  /**
+   * @notice Handles transferring funds from msg.sender to the Connext contract.
+   * @dev If using the native asset, will automatically wrap
+   * @param _assetId - The address to transfer
+   * @param _assetAmount - The specified amount to transfer. May not be the
+   * actual amount transferred (i.e. fee on transfer tokens)
+   * @param _fee - The fee amount in native asset included as part of the transaction that
+   * should not be considered for the transfer amount.
+   * @param _wrapper - The address of the wrapper for the native asset on this domain
+   * @return The assetId of the transferred asset
+   * @return The amount of the asset that was seen by the contract (may not be the specifiedAmount
+   * if the token is a fee-on-transfer token)
+   */
+  function _handleIncomingAsset(
```

Comment:
2 wrapper functions called `_handleIncomingAsset` ? And neither of them use `AssetLogic.handleIncomingAsset` ?